src/volatile_memory.rs

```diff
@@ -263,10 +286,11 @@ impl<'a> VolatileSlice<'a> {
     /// and is available for the duration of the lifetime of the new `VolatileSlice`. The caller
     /// must also guarantee that all other users of the given chunk of memory are using volatile
     /// accesses.
-    pub unsafe fn new(addr: *mut u8, size: usize) -> VolatileSlice<'a> {
+    pub unsafe fn new(addr: *mut u8, size: usize, bitmap: B) -> VolatileSlice<'a, B> {
         VolatileSlice {
             addr,
             size,
+            bitmap,
             phantom: PhantomData,
         }
     }
```

Comment:
@alexandruag I found it was needed to mark the whole slice dirty when using `VolatileSlice::as_ptr()`, as we may have write access to the  returned mutable pointer. One example is the asynchronous operations on virtio-blk queues in [Cloud Hypervisor](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/block_util/src/lib.rs#L320-L323).  I made the above change in my [own fork of your branch](https://github.com/likebreath/vm-memory/commits/bitmap_get_and_reset), and tested it with [Cloud Hypervisor](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2497). Now I have Cloud Hypervisor migrating successfully  a guest workload (with intensive virtio-blk transactions) that was interrupted during live-migration previously (without vm-memory dirty page tracking). 

I didn't send a PR directly to your branch from my fork because the above change in `VolatileSlice::as_ptr()` breaks your unit test `test_volatile_slice_dirty_tracking()`. The reason is `VolatileSlice::subslice()` uses the `as_ptr()` function and  will mark the whole slice dirty. 

For the similar reason, I think  `pub fn as_ptr()` for `VolatileRef` and `VolatileArrayRef` should also mark the referred memory dirty.

Please let me know your thoughts or questions. Thank you.

Reply:
Thanks for bringing this up, it's a very interesting discussion. Adding the semantics you propose to the `as_ptr` method(s) is definitely an option, but I think it makes the abstraction quite inflexible, because a single `as_ptr` call will mark an entire slice as dirty, no matter how large it is vs the actual access size. My preferred approach is to explicitly call something like `slice.bitmap().mark_dirty(offset, len)` in conjunction with the use of `slice.as_ptr()` when getting the pointer for write access. This adds an extra step, but allows finer grained control instead of forcing the tracking functionality to assume the entire slice is going to become dirty. Do you think this pattern is compatible with the use cases present in Cloud Hypervisor?

Reply:
@alexandruag Thanks for the detailed comments. Yes, I think so. I don't see it is problem to ask the user of `as_ptr` to explicitly set the  `bitmap` in Cloud Hypervisor. I will confirm that in my integration PR and keep you posted.

This pattern makes sense, and I think we need to make it louder/clearer about which functions need explicit bitmap tracking from their users/callers. I remember you added relevant comments to some other functions, and we can do the same to `as_ptr` functions at least. 

Also, for the same reason (to maximize the precision of dirty page tracking), do you think we need to enforce the same pattern in functions `VolatileMemory::aligned_as_mut/get_atomic_ref` (where we are marking dirty the whole slice being returned)?

Reply:
@alexandruag In case you missed it, I want to let you know that the pattern you mentioned above (handling special cases of dirty-page-tracking explicitly from the VMM) works fine with Cloud Hypervisor ([details](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2497/commits/a535a6470afd90994fec642c10ef0fab840d51ff)).

Also, to complete the live-migration support in Cloud Hypervisor, we need to have the dirty-page-tracking of the `vm-memory` crate sooner than later. So we are planning to pull your PR changes into our own fork of  the `vm-memory` crate (with the addition of `get_and_reset` function), and reply on that fork before your PR is landed in the upstream.

Meanwhile, we are more than happy to continue the discussions here and hopefully we can help with landing the PR faster. Please let me know if you have any questions about my previous comments. I will also keep you posted if we have any new findings with more experiments of live-migration in Cloud Hypervisor using your patches. Thank you.

/cc @rbradford